File: mypay.sources/mypay4-fe/projects/mypay4-fe-cittadino/src/app/services/recaptcha.service.ts
/*
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import {
  ConfigurationService, UserService, genericRetryStrategy
} from 'projects/mypay4-fe-common/src/public-api';
import { BehaviorSubject, Observable, ReplaySubject, Subject, from, of, throwError } from 'rxjs';
import { catchError, first, flatMap, retryWhen } from 'rxjs/operators';

import { Component, Injectable, NgZone, Renderer2 } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';

declare let grecaptcha: any;

@Injectable({
  providedIn: 'root'
})
export class RecaptchaService {

  private ready = new ReplaySubject<any>(1);
  private active = new BehaviorSubject<boolean>(false);

  private enabled: boolean;
  private siteKey: string;
  private siteKeyV2: string;
  private recaptchaV3Prefix: string = '';

  constructor(
    private conf: ConfigurationService,
    private userService: UserService,
    private dialog: MatDialog,
    private ngZone: NgZone,
  ) {
    this.siteKey = this.conf.getBackendProperty('recaptchaSiteKey');
    this.siteKeyV2 = this.conf.getBackendProperty('recaptchaV2SiteKey');
    this.enabled = this.siteKey ? true : false;
    console.log("recaptcha enabled: "+this.enabled);
  }

  public isEnabled():boolean {
    return this.enabled;
  }
  public isActive():boolean {
    return this.active.value;
  }
  public isActiveObs():Observable<boolean> {
    return this.active.asObservable();
  }
  public deactivate() {
    //console.log('recaptcha: deactivating');
    if(this.enabled && this.active.getValue()){
      this.active.next(false);
      //console.log('recaptcha: deactivated');
    }
  }

  public init(renderer: Renderer2) {
    if(this.enabled && typeof grecaptcha === 'undefined'){
      console.log("init recaptcha");
      if(this.conf.getFlag("recaptchaForceV2", 'false')==='true')
        this.recaptchaV3Prefix = "V2Force|";
      const script = renderer.createElement('script');
      script.src = `https://www.google.com/recaptcha/api.js?render=${this.siteKey}`;
      renderer.appendChild(document.head, script);
      this.ready.next(true);
      console.log("init recaptcha: done");
    }
    //console.log('recaptcha: activating');
    if(this.enabled)
      this.active.next(true);
  }

  public initFallback(callback: (recaptchaResponse: string)=>void){
    grecaptcha.ready(() => {
      const recaptchaV2DialogRef = this.dialog.open(RecaptchaV2Dialog);
      const recaptchaV2WidgetId = grecaptcha.render('recaptcha_fallback', {
          'sitekey' : this.siteKeyV2,
          'callback' : (response) => {
            setTimeout(()=>{
              grecaptcha.reset(recaptchaV2WidgetId);
              this.ngZone.run(() => {
                recaptchaV2DialogRef.close();
                callback(response);
              });
            });
          }
      });
    });
  }

  private submitToken(action: string):Observable<string>{
    if(this.enabled)
      return this.ready.pipe(
        flatMap(() => from(grecaptcha.execute(this.siteKey, {action: action}))),
        retryWhen(genericRetryStrategy({scalingDuration: 100})),
        catchError(error => {
          console.log('error submitting captcha', error);
          return throwError(error || 'Errore durante la verifica del codice reCaptcha');
        }),
        first()
      ) as Observable<string>;
    else
      return of('reCaptcha_disabled');
  }

  

  public submitWithRecaptchaHandling<T>(action: string,
    payloadFunAnon: (recaptchaResponse: string)=>Observable<T>,
    payloadFunLogged?: ()=>Observable<T>) {

    if(this.userService.isLogged())
      return payloadFunLogged?.() ?? of(null);
    else
      return this.submitToken(action).pipe(
        flatMap(token => payloadFunAnon(this.recaptchaV3Prefix + token)),
        catchError(error => {
          if(error?.status===471 && error?.error==='recaptcha_low_score' && this.siteKeyV2){
            console.log('falling back to recaptcha V2');
            const v2Callback = new Subject<string>();
            this.ngZone.runOutsideAngular(()=>{
              this.initFallback(recaptchaResponseV2 => {
                v2Callback.next(recaptchaResponseV2);
                v2Callback.complete();
              });
            });
            return v2Callback.asObservable().pipe(
              flatMap(token => payloadFunAnon("V2|"+token)),
            );
          } else {
            return throwError(error);
          }
        })
      );

  }

}


@Component({
  selector: 'recaptchaV2-dialog',
  templateUrl: 'recaptchaV2-dialog.html',
})
export class RecaptchaV2Dialog {
  constructor() {}
}

File: mypay.sources/mypay4-be/src/main/java/it/regioneveneto/mygov/payment/mypay4/service/RecaptchaService.java
/**
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package it.regioneveneto.mygov.payment.mypay4.service;

import it.regioneveneto.mygov.payment.mypay4.dto.RecaptchaResponseTo;
import it.regioneveneto.mygov.payment.mypay4.exception.RecaptchaFallbackException;
import it.regioneveneto.mygov.payment.mypay4.util.Utilities;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

@Service
@Slf4j
public class RecaptchaService {

  public static final String V_2 = "V2|";
  public static final String V_2_FORCE = "V2Force|";
  private final RestTemplate restTemplate;

  public RecaptchaService(RestTemplateBuilder restTemplateBuilder) {
    this.restTemplate = restTemplateBuilder.build();
  }

  @Value("${google.recaptcha.enabled}")
  private String enabled;
  @Value("${google.recaptcha.secret.key}")
  private String recaptchaSecret;
  @Value("${google.recaptcha.v2.enabled:false}")
  private String v2Enabled;
  @Value("${google.recaptcha.v2.secret.key:}")
  private String recaptchaSecretV2;
  @Value("${google.recaptcha.verify.url}")
  private String recaptchaVerifyUrl;
  @Value("${google.recaptcha.score.threshold:0.5}")
  private double scoreThreshold;

  public boolean verify(String response, String action) {
    if(!Boolean.parseBoolean(enabled))
      return true;

    boolean isV2Enabled = Boolean.parseBoolean(v2Enabled) && StringUtils.isNotBlank(recaptchaSecretV2);

    boolean isV2 = StringUtils.startsWith(response, V_2);
    boolean isV2Force = StringUtils.startsWith(response, V_2_FORCE);

    if(isV2) {
      if(!isV2Enabled) {
        log.warn("trying recaptcha V2 validation but V2 is not enabled or key missing");
        return false;
      }
      log.debug("performing fallback recaptcha validation V2");
      response = response.substring(V_2.length());
    }
    if(isV2Force){
      response = response.substring(V_2_FORCE.length());
    }

    MultiValueMap<String, String> param= new LinkedMultiValueMap<>();
    param.add("secret", isV2 ? recaptchaSecretV2 : recaptchaSecret);
    param.add("response", response);
    try {
      RecaptchaResponseTo recaptchaResponse = this.restTemplate.postForObject(recaptchaVerifyUrl, param, RecaptchaResponseTo.class);

      boolean successWithoutScoreCheck = recaptchaResponse!=null && recaptchaResponse.isSuccess() && ( isV2 || action.equals(recaptchaResponse.getAction()) );
      boolean scoreV3Check = !isV2 && recaptchaResponse!=null && recaptchaResponse.getScore() > scoreThreshold;

      //just for test: force use of recaptcha v2
      if(isV2Force) {
        log.warn("recaptcha validation V3, force use V2; originalScore[{}]", Utilities.ifNotNull(recaptchaResponse, RecaptchaResponseTo::getScore));
        scoreV3Check = false;
      }

      boolean success;
      if(isV2)
        success = successWithoutScoreCheck;
      else if(isV2Enabled && successWithoutScoreCheck && !scoreV3Check)
        throw new RecaptchaFallbackException("recaptcha low score: ["+recaptchaResponse.getScore()+"]", recaptchaResponse.getScore());
      else
        success = successWithoutScoreCheck && scoreV3Check;

      if(success)
        log.debug("recaptcha V2[{}] valid for action [{}]: [{}]", isV2, action, recaptchaResponse);
      else
        log.warn("recaptcha V2[{}] NOT valid for action[{}]: [{}]", isV2, action, recaptchaResponse);

      return success;
    }catch(RestClientException e){
      log.error("error verifying recaptcha for response "+response, e);
      return false;
    }
  }

}

